Issue 523

berinhard opened the pull request:
Closes https://github.com/model-bakers/model_bakery/issues/523

This PR introuces the `_full_clean` flag to every API to create/prepare objects. By default the flag is set to `False` to not introduce backwards incompatibilities.

berinhard committed:
Add default values to JSON field to avoid extra validation

berinhard committed:
Make sure flag triggers full clean validation before saving object

berinhard committed:
Enable flag to control full clean to recipes also

berinhard committed:
Add option to run full clean after bulk creation

berinhard committed:
Make sure that instances can be prepared with full clean

berinhard committed:
Add docs on new _full_clean boolean

berinhard committed:
Update changelog

amureki commented on model_bakery/baker.py:
```diff
@@ -916,4 +947,7 @@ def bulk_create(baker: Baker[M], quantity: int, **kwargs) -> list[M]:
                     )
 
+        if _full_clean:
+            entry.full_clean()
```
I think, this one is being called too late. We already created objects in line 918: `created_entries = manager.bulk_create(entries)`, and only now doing a validation.  This does not prevent objects from being saved to the db.

As an alternative, we could pass it in baker.prepare a bit higher in the stack:

```
entries = [
    baker.prepare(
        _full_clean=_full_clean,
        **kwargs,
    )
    for _ in range(quantity)
]
```

amureki reviewed (changes requested):
I've got a couple of comments on this PR. And I left a comment on the issue itself, as I am curious more about the context.

Many thanks for the contribution, though!

berinhard committed:
Run full clean before creating related objects during bulk creation

berinhard committed:
Make sure object and related objects aren't created on bulk if error

berinhard committed:
Fix breaking tests due to new transaction block

Since the bulk create is being done within a transaction block to prevent
objects from being created if invalid error, the number of expected queries
increases by 2 since the transaction runs the statement to register the
rollback checkpoint and the commit one.

berinhard committed:
Make sure the object does not get created if full clean on make

berinhard committed:
Update changelog

berinhard committed:
Code linter as always

berinhard commented:
Thanks for you review @amureki. I addressed your comments with my new commits :+1:

amureki commented on .envrc:
I believe, this file should not be commited, as it belongs to a local dev setup.
I think you recently added it to .gitignore, so this is solved on main branch already.

amureki commented on CHANGELOG.md:
```diff
@@ -9,10 +9,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
+- New flag `_full_clean` to run model validation at created instances (`False` by default)
 
 ### Changed
+- Bulk creation of models instances is now done within a Django transaction block
```
```suggestion
- Bulk creation now uses transaction blocks to ensure atomic rollback when `_full_clean=True` validation fails
```

amureki reviewed (commented):
Excellent fixes, thanks for moving the check and adding the test!
I left some last nitpicky notes, but in general it is good to go!

amureki commented:
Finalized in #591

Thank you for the great start that served as a base! ❤️
